### PR TITLE
Capture and return the exact string type of `extract*` arguments

### PR DESCRIPTION
**Pull request by layday**

### CODE REVIEW

**Comment by layday** [comment]
(`choices` also only accepts collections, not iterables or iterators, since `extract` calls `len` on it.)

File: src/rapidfuzz/process.pyi [hunk 1]
```diff
@@ -2,29 +2,30 @@ from typing import Any, Mapping, Tuple, Callable, Hashable, Sequence, Iterable,
 from rapidfuzz.fuzz import WRatio
 
 _StringType = Sequence[Hashable]
+_AnyStringType = TypeVar("_AnyStringType", bound=_StringType)
```
Comment (maxbachmann) [inline src/rapidfuzz/process.pyi:5]: Could you give me a brief explanation of the difference between _StringType and _AnyStringType (I am not very familiar with pyton typing)?

Comment (layday) [reply src/rapidfuzz/process.pyi:5]: The difference is that `_AnyStringType` captures the type passed to the function as an argument and returns it. So where previously you'd pass, e.g. a `str` and get `Sequence[Hashable]` out of `extract`, you can now pass a `str` _and_ get a `str` back.  The implication is that if you are manipulating the output of `extract`, other code might expect the resultant type to be `<your type>` and not just any `Sequence[Hashable]`.  A type error will also be produced if the type of `query` differs from the type of `choices` items, which wasn't the case before.

Comment (maxbachmann) [reply src/rapidfuzz/process.pyi:5]: I think this should differentiate between the type of query and choices.
```
def extractOne(
    query: _AnyStringType1,
    choices: Iterable[_AnyStringType2], *,
    scorer: Callable[..., ResultType] = WRatio,
    processor: Optional[bool] = None,
    score_cutoff: Optional[ResultType] = None,
    **kwargs: Any
) -> Tuple[_AnyStringType2, ResultType, int]: ...
```

So `extractOne(str, list[list[str]])` returns `Tuple[list[str], ...]`.

Comment (layday) [reply src/rapidfuzz/process.pyi:5]: I've put it back as `_StringType` (it doesn't need to be a type var if it only appears once in the function signature).

File: src/rapidfuzz/process.pyi [hunk 2]
```diff
@@ -1,35 +1,36 @@
...
 S1 = TypeVar("S1")
 S2 = TypeVar("S2")
 ResultType = TypeVar("ResultType", int, float)
 
 @overload
 def extractOne(
     query: _StringType,
-    choices: Iterable[_StringType], *,
+    choices: Collection[_AnyStringType], *,
```
Comment (maxbachmann) [inline src/rapidfuzz/process.pyi]: extractOne and extract_iter do not use the container length. So it should work with Iterables and not only Collections.

Comment (layday) [reply src/rapidfuzz/process.pyi]: Thanks, fixed.

**Comment by maxbachmann** [comment]
Thanks

**Pull request merged**